NodeType checking when adding PSyIR children

sergisiso opened the pull request:
In this PR I add the Statement and DataNode abstract nodes (see #723) and automatic validation checks to all nodes inserted to a PSyIR tree.

In order to do the validation, I first thought to add checks in the `addchild()` method, but quickly realized this was easy to circumvent with list operations like:
`node.children.append(node)`
`node.children = [node1, node2]`
which appear often in the code.

Putting restrictions on how to access the children list would be quite difficult (an unpythonic), so as an alternative solution I customized (by subclassing) the children list with an injected validation to each insertion method e.g.:
```
def append(self, item):
    self._validate_item(len(self), item)
    super(ChildrenList, self).append(item)
```
Another benefit of this PR is that it centralizes the validation logic to the same place and just requires a validation method (often 1 line) in each node, but the rest of the code remains unaffected (other than a few bugs that popped up with the validation).
This mean I can remove a few duplicated check across the code.

sergisiso commented:
@rupertford @arporter Since in this PR I made the unconventional design decision to subclass list, I will ask for a first review of the code before finishing the unit-tests (altough I recommend looking at `src/psyclone/tests/psyir/nodes/node_test.py`) and removing the duplicated checks that now appear at the begging of several methods.

arporter commented:
I've had a quick look through and I like it. Fundamentally `ChildrenList` doesn't alter the semantics of `list`, just augments them with checking (with a clear error message if the checks fail). The problems we've had in fparser stem from things like the `==` operator being overridden so that the behaviour is no longer what one would expect. The changes in this PR won't cause problems like that.
We should probably get @rupertford to give his (dis)approval too... :-)

rupertford commented on src/psyclone/psyGen.py:
```diff
@@ -1147,6 +1149,20 @@ def __init__(self, ast=None, children=None, parent=None):
 
+
+    @staticmethod
+    def _validate_child(possition, child):
```
position

sergisiso replied:
Fixed

rupertford commented:
I agree, nicely done. I like the DataNode and Statement subclassing too. Perhaps the textual description of valid children could go inside the check? Any way I don't disagree with the subclassing of list :-)

sergisiso commented:
In  this PR I also moved the _text_name and _colour_key attributes to be class attributes. I did this for the following reasons:
- Sometimes they needed to be defined before calling the super()__init when some error message was writing out the class name. And sometimes after super()__init, when the chain of inits was replacing the text. The new implementation is more consistent.
- Some definitions seemed out of place: they were needed because some constructors do not call the super() constructor, so the code needed to be repeated in the specialized class. Class attributes follow the class hierarchy as expected.
- By default not having a definition meant use the class name, first I thought this was a good idea but we needed to hardcode some names if the sub-classes used the same text. Having to know the needs of your sub-class specialisations does not seem a good design.
- They are constant for each class :)

sergisiso commented:
@arporter @rupertford This is ready for the first review. It has all the new functionality but I have been reserved in removing `isinstance(node, Node)` checks around the code that technically can now be removed. Should I be more aggressive with this?

arporter commented:
> 
> 
> @arporter @rupertford This is ready for the first review. It has all the new functionality but I have been reserved in removing `isinstance(node, Node)` checks around the code that technically can now be removed. Should I be more aggressive with this?

I would vote 'yes' as you know what can be safely removed. I suspect if we leave it we'll forget what we can and cannot remove. Could you point me to some example code so I can see whether I want to change my mind :-)

rupertford commented:
> > @arporter @rupertford This is ready for the first review. It has all the new functionality but I have been reserved in removing `isinstance(node, Node)` checks around the code that technically can now be removed. Should I be more aggressive with this?
> 
> I would vote 'yes' as you know what can be safely removed. I suspect if we leave it we'll forget what we can and cannot remove. Could you point me to some example code so I can see whether I want to change my mind :-)

IMO if they have been made redundant due to this PR then it makes sense to remove them in this PR. Hopefully that is not too much extra effort.

sergisiso commented:
@rupertford @arporter This is ready for review again (I hope this time for real). As suggested in the call this morning I deleted duplicated input argument node type validations but not internal consistency checks that may still be useful.

rupertford commented on src/psyclone/psyir/nodes/container.py:
```diff
@@ -55,11 +55,32 @@ class Container(Node):
 
     '''
+    # Class attributes of this node textual representations
+    _children_valid_format = "*[Container | KernelSchedule]"
```
I presume * means 0 or more? If so, I would expect the * to be at the end if we follow the usual regex format. Could you also add an explanation of the use of children_valid_format and the actual format used in the developers guide. While I am asking could you please add a section explaining what _validate_child does, what its requirements are and how it is implemented (i.e. the subclassing of list). Some of the info is already written in the PR. If you could also mention the need to add _colour_key and why as well that would be great.

sergisiso replied:
I moved * and + to the end and wrote a new developer_guide section.

rupertford commented on src/psyclone/psyir/nodes/literal.py:
```diff
@@ -92,6 +96,20 @@ def __init__(self, value, datatype, parent=None):
         self._datatype = datatype
         self._value = value
 
+    @staticmethod
```
Perhaps it would also make sense to create a LeafDataNode as well - depending on how many "<LeafNode>" classes there are? Again this assumes you don't go for the super-base-class-omatic solution.

sergisiso replied:
Look at the new implementation where the validate acts like a Leaf by default. Is this implementation ok?

rupertford replied:
Yes.

rupertford commented on src/psyclone/psyir/nodes/node.py:
```diff
@@ -95,9 +95,145 @@ def colored(text, _):
+    making use of the validation.
+
+    :param node: reference to the node where the list belongs.
+    :type node: :py:class:`psyclone.psyir.nodes.node`
```
``node`` -> ``Node``

sergisiso replied:
Done

rupertford commented on src/psyclone/psyir/nodes/node.py:
```diff
@@ -95,9 +95,145 @@ def colored(text, _):
+        item into the list.
+
+        :param int index: position where the item is inserted into.
+        :param item: object to be added into the list.
```
No type for this param

sergisiso replied:
Fixed

rupertford commented on src/psyclone/psyir/nodes/node.py:
```diff
@@ -95,9 +95,145 @@ def colored(text, _):
         return text
 
 
+class ChildrenList(list):
```
Please explain this v nice approach and what it requires when adding new nodes in the developers guide

sergisiso replied:
Added an explanation in the developers guide.

rupertford commented on src/psyclone/psyir/nodes/node.py:
```diff
@@ -95,9 +95,145 @@ def colored(text, _):
+        ''' Extends list pop method with children node validation.
+
+        :param int index: position where to insert the item.
+
```
This returns something? If so, please include a docstring.

sergisiso replied:
Done

rupertford commented on src/psyclone/psyir/nodes/node.py:
```diff
@@ -123,12 +259,16 @@ class Node(object):
     # The list of valid annotations for this Node. Populated by sub-class.
     valid_annotations = tuple()
+    # Class attributes of this node textual representations
+    _children_valid_format = "*[Node]"
```
Would it make sense to set these to None here and then you could check that they are set by subclasses - a sort of abstract attribute? Or are you happy with the default being Node and people can override if they want to (plus if they read the developers guide which will explain :-)).

sergisiso replied:
I like the `None` idea (there may be better solution for Python > 2 e.g. https://stackoverflow.com/a/53417582) but since we still support Python 2 I changed them to None for now.

rupertford replied:
I am happy with None but we already use abc in PSyclone so I don't see why we could not use it if we wanted to.

sergisiso replied:
Because what we need here is "ABC.abstractclassmethod" which was added to python 3, and is not available in python 2.

sergisiso commented on src/psyclone/dynamo0p3.py:
```diff
@@ -5602,6 +5602,9 @@ class DynHaloExchangeStart(DynHaloExchange):
     :type parent: :py:class:`psyclone.psyGen.node`
 
     '''
+    # Class attribute of this node textual representation
```
Nothing that `find src/ -type f -name '*.py' -exec sed -i 's/FROM/TO/g' {} \;` can't handle :)

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -1042,13 +1044,28 @@ class Directive(Node):
+    @staticmethod
+    def _validate_child(position, child):
+        '''
+        :param int position: a position to be validated.
```
Done

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -1042,13 +1044,28 @@ class Directive(Node):
+        '''
+        :param int position: a position to be validated.
+        :param child: a child to be validated.
+        :type child: :py:class:`psyclone.psyir.nodes.node`
```
Done (again `find`&`sed -i` save the day)

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -2274,8 +2325,11 @@ class Kern(Node):
+
     def __init__(self, parent, call, name, arguments):
-        Node.__init__(self, children=[], parent=parent)
+        super(Kern, self).__init__(self, parent=parent)
```
Node already initializes `_children` to an empty list when receives None, so both are equivalent, and the new line uses proper polymorphic `__init__` and one less argument, so I think it is better.

What I am not convinced in that Node accepts children as a parameter for top-down initialization and some of the sub-class nodes that this one (Kern) doesn't.

I think we have to decide if children is accepted in all constructors and then passed to the super constructor, or it is removed from all the constructors. But this is covered in another issue.

sergisiso commented on src/psyclone/psyir/backend/visitor.py:
```diff
@@ -111,8 +111,6 @@ def reference_node(self, node):
         :returns: the Fortran code as a string.
         :rtype: str
 
-        :raises VisitorError: if this node has children.
```
Oops, that was a mistake.

sergisiso commented on src/psyclone/nemo.py:
```diff
@@ -549,8 +544,6 @@ def __init__(self, ast, parent=None):
                       valid_loop_types=valid_loop_types)
         # Keep a ptr to the corresponding node in the AST
         self._ast = ast
-        self._text_name = "NemoImplicitLoop"
```
That was not intended, I think I deleted this at the end because CodeCov said it was not being used. In a more careful look I see the "NemoImplicitLoop" was hardcoded in the `__str__` method. I restored this _text_name and add the node_str method for proper coloring.

(Note that the other _text_name's deleted from the nemo.py file didn't really belong here)

sergisiso replied:
I seen there are some more hardcoded str methods that are not necessary, I fixed some in the code that I touched.

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -1042,13 +1044,28 @@ class Directive(Node):
+        :rtype: bool
+
+        '''
+        return position == 0 and isinstance(child, Schedule)
```
I see your point and it is a nice idea. I think we don't even need the dictionary as python eval may already be able convert the strings to the class type. But since getting a regex parser right may be tricky I think it is better to leave this to a separate PR.

Also, forcing the validation to be always done with a true regex expression could limit the PSyIR language in the future. Are we sure a Regular Grammar will cover all node types and hierarchies we want to add?

rupertford replied:
I'm certainly not sure :-). If you think this is something worth looking into then could you create a new issue and add a reference to it in the code. If not, don't worry, it was just a thought.

sergisiso replied:
I created an issue and added references to it.

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -2061,8 +2083,20 @@ def __init__(self, scalar, parent=None):
+        :rtype: bool
+
+        '''
+        # pylint: disable=unused-argument
```
I deleted this and all leaf node validations, but I added a similar comment to the generic implementation.

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -2131,8 +2170,20 @@ def __init__(self, field, check_dirty=True,
-        self._text_name = "HaloExchange"
-        self._colour_key = "HaloExchange"
+
+    @staticmethod
```
I don't like the subclassing here as leaf statements can appear in different places of the current class hierarchy and we would need multiple inheritance for it. But I made the generic implementation behave like a Leaf node. (I think in this case is it worth breaking the OO rule of expansion>>>modification since the non-leaf nodes have to always specialize its own validate method anyway)

The effect is the same, I could delete all duplicated Leaf node validations.

sergisiso commented on src/psyclone/psyir/nodes/codeblock.py:
```diff
@@ -38,11 +38,12 @@
 
 
-class CodeBlock(Node):
+class CodeBlock(Statement, DataNode):
```
In inherits for both because for the child validation a CodeBlock can be found in the place of both, a Statement or a DataNode. Being child of Node wouldn't be an option. The alternative would be to create 2 classes: `CodeBlockStatement(Statement)` and `CodeBlockDataNode(DataNode)`.

I remember you also had to differentiate them in the back-end and end-up using the Structure attribute and enumeration inside the class. Maybe going to the 2 different classes system will be a better solution altogether?

I will mention something along this lines in the developer guide.

rupertford replied:
Yes I meant developers guide when I said reference guide :-)

rupertford replied:
Sorry if I missed it but I could not see any text talking about CodeBlock(Statement, DataNode) vs CodeBlockStatement(Statement) and CodeBlockDataNode(DataNode) in the developers guide.

rupertford replied:
If it is a better solution then we should make an issue and reference it. If it is not clear then it is fine to discuss in the developers guide.

sergisiso replied:
> I will mention something along this lines in the developer guide.

When I wrote this comment in future tense I thought that if I didn't do it immediately I would have a 90% change to forget about it. And I didn't do it immediately :)

I think of it more as an alternative strategy, so I just changed a bit the CodeBlock section of the developer guide to explain this. (I believe it had some exceptions regarding Directives that are not true anymore, so I also changed that part.)

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -3102,6 +3189,10 @@ class BuiltIn(Kern):
     does not have to provide an implementation).
     '''
+    # Class attributes of this node textual representations
+    _text_name = "BuiltIn"
```
They can't be the same as often the subclasses redefine the name but not the colour. e.g. BinaryOperation, UnaryOperation redefine the name they show in the `view()` but not the colour, which is inherited from Operation.

Using the same key will mean we need to add a new entry to the COLOUR_MAP for each new class. Which is something we conveniently avoid to to in the APIs with the current implementation.

One alternative would be to get rid of the COLOUR_MAP altogether and use colours directly `_colour_key="green"` but in any case this should be discussed outside the scope of this PR.

sergisiso commented:
@rupertford This is ready for the next review. I addressed the inline comments, the biggest changes are the removal of leaf node validations and the expansion of the developers guide PSyIR (I see git diff has some troubles, what I did is to move the symbol and types info in a new page (untouched text) and added a new introduction and how to create PSyIR node sections in the PSyIR page. I also replaced inline autodocs with links to the reference guide.)

rupertford commented on doc/developer_guide/psyir.rst:
```diff
@@ -37,72 +37,147 @@ The PSyclone Internal Representation (PSyIR)
-where ``RealType`` subclasses ``ScalarType``. It may be that the
-latter would have provided a better interface, but both approaches have
-the same functionality.
+.. note:: For convinience, the PSyIR children validation is performed
```
convenience

sergisiso replied:
Fixed

rupertford commented on doc/developer_guide/psyir.rst:
```diff
@@ -300,10 +365,9 @@ for CodeBlocks. However, the purpose of the PSyIR is to capture code
 concepts that are relevant for performance, not all aspects of a code,
 therefore it is likely that that CodeBlocks will continue to be an
```
Not your code but we have "that that". Could you please fix this.

sergisiso replied:
Fixed

rupertford replied:
Thanks

rupertford commented on src/psyclone/tests/psyir/transformations/transformations_test.py:
```diff
@@ -40,7 +40,8 @@
 from __future__ import absolute_import, print_function
 import pytest
-from psyclone.psyir.nodes import Literal, Loop, Node, Reference, Schedule
+from psyclone.psyir.nodes import Literal, Loop, Node, Reference, Schedule, \
```
add modified by you please.

sergisiso replied:
done

sergisiso commented on doc/developer_guide/psyir.rst:
```diff
@@ -37,72 +37,147 @@ The PSyclone Internal Representation (PSyIR)
+See the full documentation of the Node class in the
+`Node reference guide <https://psyclone-ref.readthedocs.io/en/latest/
+autogenerated/psyclone.psyir.nodes.html#psyclone.psyir.nodes.Node>`_.
+The remaining of this page contains information about :ref:`newnodes-label`
```
Deleted

sergisiso commented on doc/developer_guide/psyir.rst:
```diff
@@ -37,72 +37,147 @@ The PSyclone Internal Representation (PSyIR)
+The remaining of this page contains information about :ref:`newnodes-label`
+and and :ref:`nodesinfo-label`.
+
+More developer information about the type system and symbols and how PSyIR
```
Deleted

sergisiso commented on doc/developer_guide/psyir.rst:
```diff
@@ -37,72 +37,147 @@ The PSyclone Internal Representation (PSyIR)
+
+More developer information about the type system and symbols and how PSyIR
+can be transformed back to a particular language using the back-ends
+(Writers) are provided in the next sections of the developer guide.
```
done

sergisiso commented on doc/developer_guide/psyir.rst:
```diff
@@ -37,72 +37,147 @@ The PSyclone Internal Representation (PSyIR)
-PSyIR DataTypes currently support Scalar and Array types via the
-``ScalarType`` and ``ArrayType`` classes. However, they are designed
-to be easily extended. For example, a structure could be created thus:
+Intermediate Representation that PSyclone uses to represent the PSy (Parallel
```
My idea here was to first 'set the stage' with which are the design requirements of PSyIR (expressive, mutable, extensible, heterogeneous, normalized), so that the design decisions below can be linked to these requirements. But I agree it looks good as a more generic introduction to PSyIR.

sergisiso commented on doc/developer_guide/psyir.rst:
```diff
@@ -37,72 +37,147 @@ The PSyclone Internal Representation (PSyIR)
+=============================
+In order to create a new PSyIR node, either for adding a new core PSyIR node
+or to extend the functionality in one of the PSyclone APIs, it is mandatory
+to steps perform the following steps: 
```
Oops, removed

sergisiso commented on doc/developer_guide/psyir.rst:
```diff
@@ -37,72 +37,147 @@ The PSyclone Internal Representation (PSyIR)
+to steps perform the following steps: 
+
+1. The new node must inherit from ``psyclone.psyir.nodes.Node`` or one of its
+   sub-classes. Note that ``Node`` won't be accepted as child anywhere in the
```
done

sergisiso commented on doc/developer_guide/psyir.rst:
```diff
@@ -37,72 +37,147 @@ The PSyclone Internal Representation (PSyIR)
+
+1. The new node must inherit from ``psyclone.psyir.nodes.Node`` or one of its
+   sub-classes. Note that ``Node`` won't be accepted as child anywhere in the
+   tree, if any of the more specialised classes fit for the new node, a good
```
Done

sergisiso commented on doc/developer_guide/psyir.rst:
```diff
@@ -37,72 +37,147 @@ The PSyclone Internal Representation (PSyIR)
+
+1. The new node must inherit from ``psyclone.psyir.nodes.Node`` or one of its
+   sub-classes. Note that ``Node`` won't be accepted as child anywhere in the
+   tree, if any of the more specialised classes fit for the new node, a good
```
Done

sergisiso commented on doc/developer_guide/psyir.rst:
```diff
@@ -37,72 +37,147 @@ The PSyclone Internal Representation (PSyIR)
+   Schedule) or ``psyclone.psyir.nodes.DataNode`` (which will be accepted
+   anywhere that the node can be evaluated to a data element). 
+
+2. Set the ``_text_name`` and the ``_color_key`` class string attributes. This
```
Done

sergisiso commented on doc/developer_guide/psyir.rst:
```diff
@@ -37,72 +37,147 @@ The PSyclone Internal Representation (PSyIR)
+   anywhere that the node can be evaluated to a data element). 
+
+2. Set the ``_text_name`` and the ``_color_key`` class string attributes. This
+   attributes will provide a standard behaviour for the ``__str__`` and
```
done

sergisiso commented on doc/developer_guide/psyir.rst:
```diff
@@ -37,72 +37,147 @@ The PSyclone Internal Representation (PSyIR)
+   attributes will provide a standard behaviour for the ``__str__`` and
+   ``view()`` methods.
+
+3. Set the ``_children_valid_format`` class string attribute and specialize
```
Argg, sorry, I replaced a few of them throughout the file now.

rupertford replied:
Yay, you went with the UK spelling :-)

sergisiso commented on doc/developer_guide/psyir.rst:
```diff
@@ -37,72 +37,147 @@ The PSyclone Internal Representation (PSyIR)
+
+For example, if we want to create a node that can be found anywhere where a
+statement is valid, and in turn it accepts one and only one DataNode as a
+children, we would wrote something like:
```
fixed

sergisiso commented on doc/developer_guide/psyir.rst:
```diff
@@ -37,72 +37,147 @@ The PSyclone Internal Representation (PSyIR)
+
+For example, if we want to create a node that can be found anywhere where a
+statement is valid, and in turn it accepts one and only one DataNode as a
+children, we would wrote something like:
```
fixed

sergisiso commented on doc/developer_guide/psyir.rst:
```diff
@@ -37,72 +37,147 @@ The PSyclone Internal Representation (PSyIR)
-It was decided to include datatype intrinsic as an attribute of ScalarType
-rather than subclassing. So, for example, a 4 byte real scalar is
-defined like this
+This description already provide the basic PSyIR functionality and the node
```
fixed

sergisiso commented on doc/developer_guide/psyir.rst:
```diff
@@ -37,72 +37,147 @@ The PSyclone Internal Representation (PSyIR)
-=====
+    To achieve this, we sub-classed the Python list and redefined all
+    methods that modify the list by calling first the PSyIR provided
+    validation method and if accepted, then continue calling the associated
```
done

sergisiso commented on doc/developer_guide/psyir.rst:
```diff
@@ -247,38 +314,36 @@ instances of `BinaryOperation`. (For the particular code fragment
-.. autoclass:: psyclone.psyir.nodes.UnaryOperation.Operator
-   :members:
-   :undoc-members:
+- and those having more than two by `psyclone.psyir.nodes.NaryOperation`
```
You are right, what about "those having more than two or a variable number of operands"

sergisiso commented on src/psyclone/psyir/nodes/datanode.py:
```diff
@@ -0,0 +1,46 @@
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Authors S. Siso, STFC Daresbury Lab
```
Fixed (also in statement.py)

sergisiso commented on src/psyclone/psyir/nodes/node.py:
```diff
@@ -95,9 +95,150 @@ def colored(text, _):
+        :param int index: position of the item that is popped out, if not \
+            given, the last element is popped out.
+
+        :returns: The last value or the given index value from the list.
```
fixed

sergisiso commented on src/psyclone/psyir/nodes/node.py:
```diff
@@ -582,7 +752,23 @@ def children(self):
-        self._children = my_children
+        ''' Set a new children list.
+
+        :param my_children: New list of children.
```
fixed

sergisiso commented on doc/developer_guide/psyir.rst:
```diff
@@ -37,72 +37,147 @@ The PSyclone Internal Representation (PSyIR)
+    - ``_children_valid_format`` is the textual representation that will be
+      used in error messages. It is expressed using a regex-like expression.
+
+    - ``_validate_child(position, child)`` returns a boolean deciding if the
```
done

sergisiso commented on examples/psyir/newnode.py:
```diff
@@ -0,0 +1,83 @@
+# -----------------------------------------------------------------------------
```
I moved it to `doc/developer_guide/code_snippets` and sphinx imports it, but it is not being executed (tested) and won't warn if its out-of-date (could be done in another issue?). Still I think it will potentially make the documentation easier to update in the future if relevant interfaces change.

sergisiso commented on examples/psyir/README.md:
```diff
@@ -31,7 +31,7 @@ LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 POSSIBILITY OF SUCH DAMAGE.
 
-Author R. W. Ford, STFC Daresbury Lab
+Authors: R. W. Ford, S. Siso STFC Daresbury Lab
```
Not anymore, I reverted the changes in this file

sergisiso commented on src/psyclone/psyir/nodes/extract_node.py:
```diff
@@ -70,11 +70,13 @@ class ExtractNode(PSyDataNode):
     :type options: dictionary of string:values or None
 
     '''
+    # Textual description of the node.
```
Done

sergisiso commented on src/psyclone/psyir/nodes/profile_node.py:
```diff
@@ -63,14 +63,13 @@ class ProfileNode(PSyDataNode):
         identify a region unless aggregate information is required.
 
     '''
+    _text_name = "Profile"
```
Done

sergisiso commented on src/psyclone/psyir/nodes/psy_data_node.py:
```diff
@@ -43,12 +43,13 @@
 from psyclone.errors import InternalError
 from psyclone.f2pygen import CallGen, TypeDeclGen, UseGen
-from psyclone.psyir.nodes import Node
+from psyclone.psyir.nodes.statement import Statement
```
Done

sergisiso commented on src/psyclone/psyir/nodes/ranges.py:
```diff
@@ -37,6 +37,7 @@
 ''' Module containing the definition of the Range node. '''
 
 from psyclone.psyir.nodes import Node, Literal
+from psyclone.psyir.nodes.datanode import DataNode
```
Done

sergisiso commented on src/psyclone/tests/nemo/transformations/openmp/openmp_test.py:
```diff
@@ -224,7 +224,7 @@ def test_omp_do_update():
     # Make the schedule invalid by adding a second child to the
     # OMPParallelDoDirective
-    directive.children.append(new_sched[0].loop_body[3])
+    directive.dir_body.children.append(new_sched[0].loop_body[3])
```
Done

sergisiso commented on src/psyclone/tests/psyir/backend/fortran_test.py:
```diff
@@ -523,35 +523,14 @@ def test_fw_exception(fort_writer):
-        "end subroutine tmp\n"
-        "end module test")
-    schedule = create_schedule(code, "tmp")
+    # 'unsupported' should be a node that neither its generic classes nor
```
done

sergisiso commented on src/psyclone/tests/psyir/backend/sir_test.py:
```diff
@@ -262,7 +262,11 @@ def test_sirwriter_node_1(parser):
     # pylint: disable=abstract-method
     class Unsupported(Node):
-        '''A PSyIR node that will not be supported by the SIR writer.'''
+        '''A PSyIR node that will not be supported by the SIR writer but
```
done

sergisiso commented on src/psyclone/tests/psyir/backend/visitor_test.py:
```diff
@@ -40,7 +40,8 @@
 import pytest
 from psyclone.psyir.backend.visitor import PSyIRVisitor, VisitorError
-from psyclone.psyir.nodes import Node
+from psyclone.psyir.nodes import Node, Reference, Array
```
done

sergisiso commented on src/psyclone/tests/psyir/frontend/where_handler_test.py:
```diff
@@ -146,8 +146,8 @@ def test_where_array_notation_rank():
     my_array = Array.create(DataSymbol("my_array", array_type),
-                            [Range(Literal("1", INTEGER_TYPE),
-                                   Literal("10", INTEGER_TYPE))])
+                            [Range.create(Literal("1", INTEGER_TYPE),
```
done

sergisiso commented on src/psyclone/tests/psyir/nodes/psy_data_node_test.py:
```diff
@@ -41,8 +41,9 @@
 
-from psyclone.errors import InternalError
-from psyclone.psyir.nodes import Node, PSyDataNode, Schedule
+from psyclone.errors import InternalError, GenerationError
```
done

sergisiso commented on src/psyclone/tests/psyir/nodes/test_range.py:
```diff
@@ -41,44 +41,37 @@
     INTEGER_SINGLE_TYPE, REAL_SINGLE_TYPE
 from psyclone.psyir.nodes import Range, Literal, Reference, Node
-from psyclone.psyGen import InternalError
+from psyclone.errors import InternalError, GenerationError
```
done

sergisiso commented on doc/developer_guide/psyir.rst:
```diff
@@ -37,72 +37,147 @@ The PSyclone Internal Representation (PSyIR)
-PSyIR DataTypes currently support Scalar and Array types via the
-``ScalarType`` and ``ArrayType`` classes. However, they are designed
-to be easily extended. For example, a structure could be created thus:
+Intermediate Representation that PSyclone uses to represent the PSy (Parallel
```
Isn't this already the third bullet point regarding extensibility? Or do you mean explain the difference between PSy-layer and kernel? Which in this case I am not sure it needs to be here as PSyIR just offers the options. It would be in a PSy-layer developers guide page that explains why it is implemented the way it is (which we don't have).

rupertford replied:
Yes, I take your point. I think the problem is that I was thinking of the LFRic, GOcean and NEMO-specific extentions to PSyIR as being the PSyIR as that was what the PSyIR was until you added the kernel support. My confusion is partly down to you saying that PSyclone uses the PSyIR for the PSy-layer (but most of the PSy-layer is based on PSyIR domain-specific extensions). Perhaps you could add a sentence to the "It is extensible" bullet point saying that PSyclone extends it for the DSLs it supports?

sergisiso commented on doc/developer_guide/psyir.rst:
```diff
@@ -37,72 +37,147 @@ The PSyclone Internal Representation (PSyIR)
+    error messages. Alternatively we could create an implementation where
+    the textual representation is parsed and the validation method is
+    generated automatically, hence avoiding the duplication. This will be
+    considered in the future.)
```
I added a reference to a new issue.

sergisiso commented on doc/developer_guide/psyir.rst:
```diff
@@ -37,72 +37,147 @@ The PSyclone Internal Representation (PSyIR)
+   of the new node:
+
+    - ``_children_valid_format`` is the textual representation that will be
+      used in error messages. It is expressed using a regex-like expression.
```
That is way more formal that I initially intended but I gave it a try.

rupertford replied:
Much better thanks.

sergisiso commented:
@rupertford PR ready for next review

rupertford commented on src/psyclone/tests/psyir/nodes/psy_data_node_test.py:
```diff
@@ -41,8 +41,9 @@
-from psyclone.errors import InternalError
-from psyclone.psyir.nodes import Node, PSyDataNode, Schedule
+from psyclone.errors import InternalError, GenerationError
+from psyclone.psyir.nodes import Node, PSyDataNode, Schedule, Return
```
pylint says Node is not used.

sergisiso replied:
deleted

sergisiso commented on doc/developer_guide/psyir.rst:
```diff
@@ -36,73 +36,153 @@
 
+More information about the type system and symbols and how PSyIR
+can be transformed back to a particular language using the back-ends
+(Writers) are provided in the following sections of the developer guide.
```
done

sergisiso commented on doc/developer_guide/psyir.rst:
```diff
@@ -36,73 +36,153 @@
+1. The new node must inherit from ``psyclone.psyir.nodes.Node`` or one of its
+   sub-classes. Note that ``Node`` won't be accepted as a child anywhere in
+   the tree. It may be appropriate to specialise one of the existing
+   subclasses of Node. A good starting point could be to sub-class
```
done

sergisiso commented on doc/developer_guide/psyir.rst:
```diff
@@ -36,73 +36,153 @@
+1. The new node must inherit from ``psyclone.psyir.nodes.Node`` or one of its
+   sub-classes. Note that ``Node`` won't be accepted as a child anywhere in
+   the tree. It may be appropriate to specialise one of the existing
+   subclasses of Node. A good starting point could be to sub-class
```
done

sergisiso commented on doc/developer_guide/psyir.rst:
```diff
@@ -36,73 +36,153 @@
+      whether the given child is a valid component for the given position.
 
-If we were to subclass, it would have looked something like this:
+    (Note that both attributes describe the same thing, and currently, it is
```
done

sergisiso commented on doc/developer_guide/psyir.rst:
```diff
@@ -36,73 +36,153 @@
+child, we would write something like:
+
+
+    .. literalinclude:: code_snippets/newnode.py
```
No, it is a standard Sphinx markup directive.

sergisiso commented on doc/developer_guide/psyir.rst:
```diff
@@ -36,73 +36,153 @@
 
-Nodes
-=====
+For a full list of method available in any PSyIR node see the
```
done

sergisiso commented on doc/developer_guide/psyir.rst:
```diff
@@ -313,406 +382,53 @@ PSyIR tree. In theory, language interfaces could be written between
+evaluates to a DataNode. To make this possible CodeBlock is a subclass
+of both: Statement and DataNode. However, in certain situations we
+still need to differentiate which one it is, for instance the Fortran
+back-end needs this information as expressions do not need indentation
```
done

sergisiso commented on doc/developer_guide/psyir.rst:
```diff
@@ -313,406 +382,53 @@ PSyIR tree. In theory, language interfaces could be written between
+This logic works for existing PSyIR nodes and relies on any future PSyIR
+nodes being constructed so this continues to be true. Another solution
+would be to have two different nodes: StatementsCodeBlock which subclasses
+Statement, and DataCodeBlock which subclasses DataNode. We have chose the
```
done

sergisiso commented on doc/developer_guide/code_snippets/newnode.py:
```diff
@@ -0,0 +1,82 @@
+# -----------------------------------------------------------------------------
```
oops fixed

sergisiso commented on src/psyclone/psyir/frontend/fparser2.py:
```diff
@@ -44,7 +44,7 @@
 from psyclone.psyir.nodes import UnaryOperation, BinaryOperation, \
     NaryOperation, Schedule, CodeBlock, IfBlock, Reference, Literal, Loop, \
-    Container, Assignment, Return, Array, Node, Range
+    Container, Assignment, Return, Array, Node, Range, Statement, Assignment
```
removed

sergisiso commented on src/psyclone/psyir/nodes/__init__.py:
```diff
@@ -54,11 +54,15 @@
 # this package e.g. 'from psyclone.psyir.nodes import Literal'
 __all__ = [
         'Node',
+        'DataNode',
```
Neither my pylint nor my pycodestyle find any issue with it. What does it say?

rupertford replied:
Don't worry then. I tend to use older versions of the tools :-)
However, for the record it says ...
```
C: 63, 0: Wrong hanging indentation (remove 4 spaces).
        'Node',
    |   ^ (bad-continuation)
C: 64, 0: Wrong hanging indentation (remove 4 spaces).
        'DataNode',
    |   ^ (bad-continuation)
...
```

sergisiso commented on src/psyclone/psyir/nodes/loop.py:
```diff
@@ -39,12 +39,14 @@
 ''' This module contains the Loop node implementation.'''
 
 from psyclone.psyir.nodes.node import Node
```
deleted

sergisiso commented on src/psyclone/psyir/nodes/operation.py:
```diff
@@ -43,11 +43,12 @@
 from enum import Enum
 import six
 from psyclone.psyir.nodes.node import Node
```
deleted

sergisiso commented on src/psyclone/tests/psyir/backend/visitor_test.py:
```diff
@@ -210,17 +211,20 @@ def testnode2_node(self, _):
 
         '''
+        @staticmethod
+        def _validate_child(_, node):
```
fixed

sergisiso commented on src/psyclone/psyir/nodes/psy_data_node.py:
```diff
@@ -43,12 +43,13 @@
 from psyclone.f2pygen import CallGen, TypeDeclGen, UseGen
-from psyclone.psyir.nodes import Node
+from psyclone.psyir.nodes.statement import Statement
+from psyclone.psyir.nodes.schedule import Schedule
```
deleted

sergisiso commented on src/psyclone/tests/psyir/frontend/fparser2_test.py:
```diff
@@ -1301,7 +1291,7 @@ def test_handling_assignment_stmt():
     fparser2assignment = Execution_Part.match(reader)[0][0]
 
-    fake_parent = Node()
+    fake_parent = Schedule()
```
done

sergisiso commented on src/psyclone/tests/psyir/frontend/fparser2_test.py:
```diff
@@ -2229,42 +2222,43 @@ def test_handling_binaryopbase():
     ''' Test that fparser2 UnaryOpBase is converted to the expected PSyIR
     tree structure.
     '''
     from fparser.two.Fortran2003 import UnaryOpBase
```
deleted

sergisiso commented on src/psyclone/tests/psyir/nodes/node_test.py:
```diff
@@ -43,10 +43,12 @@
+from psyclone.psyir.symbols import DataSymbol, SymbolError, \
+    INTEGER_TYPE, REAL_TYPE, SymbolTable, ContainerSymbol, \
+    UnresolvedInterface, ScalarType, DeferredType
 from psyclone.psyGen import PSyFactory, OMPDoDirective, Kern, KernelSchedule
```
deleted

sergisiso commented:
@rupertford Ready for next review.

rupertford commented on doc/developer_guide/psyir.rst:
```diff
@@ -36,73 +36,155 @@
-           :type type_list: list of :py:class:`psyclone.psyir.symbols.DataType`
+More information about the type system and symbols and how PSyIR
+can be transformed back to a particular language using the back-ends
+(Writers) are provided in the following sections of this guide.
```
are -> is

sergisiso commented:
@rupertford Brought to master and fixed remaining issue.
